rfc/004-feature-template.md

```diff
@@ -0,0 +1,51 @@
+
+Moving forward we want to use badges to indicate the satus of a table resource (alpha, beta GA, deprecated), weather we recommend using a certain resource, as well as using them to indicate if a column in a table is a primary key or partition column.
+
+## Motivation
+
+In order to best use badges we want to add sentiment and category fields to the entity to help decide things like the presentation of the badge (color, placement) and how to prioritize resources based on what badges they have. This makes the Tag and Badge entities different, as Badge requires more information now and will be used in a very different way from Tag. Separating the two now will likely lead to less tech debt in the future when other features are implemented that impact the way we ingest and use badges. Additionally, this change will help clarify the difference in use and purpose between tags and badges throughtout Amundsen. This feature will also guarantee badges are not ingested without having sentiment and category.
+
+
+## Guide-level Explanation (aka Product Details)
+
+A badge is a predetermined indicator which can be added by a table owner in order to offer information about the status of a table resource (alpha, beta GA, deprecated), whether they recommend using a certain resource, or to indicate if a column in a table is a primary key or partition column. 
+Using badges to boost results in searches and to filter resources when searching will help Amundsen users find the resources they need easier, as well as help them discern which resources are most trustworthy.
+
+## Reference-level Explanation (aka Technical Details)
+
+The implementation will require:
+- Add methods to the metadata service API to add, remove and, get badges as currently the service uses the methods for tags.
+- Update badge ingestion docs to reflect new API call and parameters.
+- Make updates to search, databuilder and frontend service to add functionality to use Badge instead of Tag.
+- Handle old badges:
+    - Keep the existing functionality and add new functionality for surfacing new badge nodes, can eventually remove old functionality
+    - Migration code to move all Tags with type “badge” to Badge nodes with default ‘sentiment’ and ‘category’ values of ‘Neutral’ and ‘Table status’ respectively. (This might only be possible down the line when we deprecate old functionality)
+
+
+## Drawbacks
+
+- Takes more effort and time to implement than the alternative of keeping them together.
+- Duplicate code for some functionality between BadgeAPI and TagAPI.
+
+## Alternatives
+
+- Keep the current functionality and simply add two optional parameters fo sentiment and category to Tag.
+
+## Unresolved questions
```

Comment:
What is the impact of changing the badges field int ables/dashboards to contain
`badge = {
    'badge_name': fields.String,
    'category': fields.String,
    'badge_type': fields.String,
}`
rather than having it be like tags (tag_type, name)?

Reply:
@ttannis where in the frontend did you say that metadata is gotten? I am looking at https://github.com/amundsen-io/amundsenfrontendlibrary/blob/ccfd2d6b82957fef347e956b243e4048c191fc0d/amundsen_application/api/metadata/v0.py and see that there is no mention of badges, only tags.

Reply:
For more context making a request like ` curl -X GET curl -X GET http://0.0.0.0:5002/"hive://gold.events/event_client_ux"/`
with what I have currently implemented on local ([PR](https://github.com/amundsen-io/amundsenmetadatalibrary/pull/176)) the response would include both fields badges and new_badges:
```   "badges": [
        {
            "tag_type": "badge",
            "tag_name": "allison"
        }
    ],```

```
    "new_badges": [
        {
            "category": "table_status",
            "badge_name": "test_dash",
            "badge_type": "neutral"
        }
    ],
```

@feng-tao suggested that we only have badges which has all badges shown according to the new_badge model. I would like to know how this change would impact other services in Amunden?

Reply:
I see. The work sounds like it would be as follows on FE:
1. Return all badges as `badges` according to the new badge model.
2. For backwards compatibility in the community, some folks could still be using Tags of `tag_type: "badge"` and we would have to account for their case. As a result we would have to include temporary logic in all of the `marshall_<some-resource>` methods in https://github.com/amundsen-io/amundsenfrontendlibrary/blob/master/amundsen_application/api/utils/metadata_utils.py. That helper method would have to look for objects of the old shape, and replace them with objects of the new shape.
3. Then the React application can write UI logic assuming the new shape and the new shape only. This would also include updating our `Tag` type interface and creating a new one for the badge object as well so that we can leverage TypeScript as usual to catch any errors when operating on the objects. 
4. Once we confirm we can officially deprecate `tag_type: "badge"` , then the code introduced in Step 2 can be deleted.

There could be some edge cases I've missed on regards to perhaps filtering support and the search service, but I'm confident we'll be able to deal with them and will advise along the way.

Reply:
However this does mean that until we deprecate Tags of `tag_type: "badge"`, apis that need badge information will also have to return tags so that FE can pull out and convert all the `tag_type: "badge"`

Reply:
We could put those in a legacy_badge field or just add logic that converts all of them to the new badge functionality with default values for category and badge_type like 'table_status' and 'neutral' respectively.

Reply:
I wonder whether we should just do a quick check on the community and see if anyone is using badge and need to have this support. Also we could always cut a break change release between all three services and document it in a doc (previously we have done sth similar when we update the model in https://github.com/amundsen-io/amundsen/blob/master/docs/faq.md#how-to-migrate-from-amundsen-1x---2x)

The hack that does the conversion in FE  is sth I would like to see if we could avoid... WDYT @allisonsuarez @ttannis ?

Reply:
Yeah I would really like to know what the current use cases for the community are so we can update accordingly and clean up anything that is not being used. I am nut sure what you mean by "cut" a break change @feng-tao

Reply:
oh, I meant cut a release version for all three services that have the breaking change.

Reply:
^ we could. But I also thought that something that would be frustrating for others is creating a new major version for individual features. There are several other items in our code base that would also benefit from being included in a planned major release, where we dedicate time to properly fixing backwards incompatible changes, deprecating configurations, etc. 

While these helper methods could seem hacky, I've seen it used elsewhere at a previous job. What made it work well and feel less hacky is that these things were well organized into separate "migration methods", and the official release process for major releases included a task to address removing the "migration methods" for what features were scheduled to be deprecated in that release. 

Overall though, I'm fine to go with whatever option the majority of maintainers support. I acknowledge that there's a chance we could get off easy this time around if the community isn't currently heavily using Tags of `tag_type: "badge"`.